File: src/devkit/plugins/go.py
"""Go plugin — latest stable toolchain from go.dev."""

from __future__ import annotations

import shutil
from pathlib import Path

from devkit.download import download_json_value, install_archive_from_url
from devkit.platform import HostOS, cpu_arch, current_os, is_windows
from devkit.plugin import EnvSpec, InstallContext, InstallResult, Plugin
from devkit.plugin_utils import binary_status
from devkit.progress import download_progress

_DL_JSON = "https://go.dev/dl/?mode=json"
MARKER = ".devkit-go"


def _go_os_arch() -> tuple[str, str]:
    host = current_os()
    arch = cpu_arch()
    if host == HostOS.WINDOWS:
        goos = "windows"
    elif host == HostOS.MACOS:
        goos = "darwin"
    elif host == HostOS.LINUX:
        goos = "linux"
    else:
        raise RuntimeError(f"Go is not supported on this OS: {host.value}")
    goarch = "arm64" if arch == "aarch64" else "amd64"
    if arch not in {"x64", "aarch64"}:
        raise RuntimeError(f"Unsupported arch for Go: {arch}")
    return goos, goarch


def resolve_go_download() -> tuple[str, str]:
    """Return ``(download_url, version)`` for the latest stable Go archive."""
    data = download_json_value(_DL_JSON)
    if not isinstance(data, list):
        raise TypeError("Unexpected go.dev/dl JSON")
    goos, goarch = _go_os_arch()
    for entry in data:
        if not isinstance(entry, dict) or not entry.get("stable"):
            continue
        version = str(entry.get("version") or "")
        for f in entry.get("files") or []:
            if not isinstance(f, dict) or f.get("kind") != "archive":
                continue
            if f.get("os") == goos and f.get("arch") == goarch:
                name = str(f.get("filename") or "")
                return f"https://go.dev/dl/{name}", version
    raise RuntimeError(f"No stable Go archive for {goos}/{goarch}")


def _go_bin(ctx: InstallContext) -> Path:
    name = "go.exe" if is_windows() else "go"
    return ctx.install_dir / "bin" / name


class GoPlugin(Plugin):
    id = "go"
    name = "Go"
    description = "Download latest stable Go toolchain and set GOROOT / PATH."

    def status(self, ctx: InstallContext):
        return binary_status(
            _go_bin(ctx),
            ctx.install_dir,
            missing_detail="Install dir exists but go binary is missing",
        )

    def install(self, ctx: InstallContext) -> InstallResult:
        url, version = resolve_go_download()
        print(f"Go {version}")
        print(f"URL: {url}")
        progress = download_progress("Downloading Go")
        # Archive root is go/; strip so bin/ lands in install_dir.
        install_archive_from_url(url, ctx.install_dir, strip_top_level=True, progress=progress)
        progress.done()
        if not _go_bin(ctx).is_file():
            raise RuntimeError(f"Go extracted but binary not found at {_go_bin(ctx)}")
        (ctx.install_dir / MARKER).write_text(version + "\n", encoding="utf-8")
        return InstallResult(ctx.install_dir, message=f"Go {version} installed at {ctx.install_dir}")

    def uninstall(self, ctx: InstallContext) -> None:
        if ctx.install_dir.exists():
            shutil.rmtree(ctx.install_dir)

    def env_spec(self, ctx: InstallContext) -> EnvSpec:
        root = str(ctx.install_dir.resolve())
        return EnvSpec(paths=[ctx.install_dir / "bin"], vars={"GOROOT": root, "GO_HOME": root})

File: src/plugins/go.rs
//! Go plugin — latest stable toolchain from go.dev.

use crate::download::{download_json_value, install_archive_from_url};
use crate::platform::{cpu_arch, current_os, is_windows, HostOS};
use crate::plugin::{EnvSpec, InstallContext, InstallResult, Plugin};
use crate::plugin_utils::binary_status;
use crate::progress::download_progress;
use anyhow::{bail, Result};
use std::path::PathBuf;

const DL_JSON: &str = "https://go.dev/dl/?mode=json";
const MARKER: &str = ".devkit-go";

fn go_os_arch() -> Result<(&'static str, &'static str)> {
    let host = current_os();
    let goos = match host {
        HostOS::Windows => "windows",
        HostOS::MacOS => "darwin",
        HostOS::Linux => "linux",
        HostOS::Other => bail!("Go is not supported on this OS: other"),
    };
    let arch = cpu_arch();
    if arch != "x64" && arch != "aarch64" {
        bail!("Unsupported arch for Go: {arch}");
    }
    let goarch = if arch == "aarch64" { "arm64" } else { "amd64" };
    Ok((goos, goarch))
}

/// Return `(download_url, version)` for the latest stable Go archive.
fn resolve_go_download() -> Result<(String, String)> {
    let data = download_json_value(DL_JSON)?;
    let entries = data
        .as_array()
        .ok_or_else(|| anyhow::anyhow!("Unexpected go.dev/dl JSON"))?;
    let (goos, goarch) = go_os_arch()?;
    for entry in entries {
        let is_stable = entry
            .get("stable")
            .and_then(|v| v.as_bool())
            .unwrap_or(false);
        if !entry.is_object() || !is_stable {
            continue;
        }
        let version = entry
            .get("version")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .to_string();
        let files = entry.get("files").and_then(|v| v.as_array());
        if let Some(files) = files {
            for f in files {
                if f.get("kind").and_then(|v| v.as_str()) != Some("archive") {
                    continue;
                }
                let f_os = f.get("os").and_then(|v| v.as_str()).unwrap_or("");
                let f_arch = f.get("arch").and_then(|v| v.as_str()).unwrap_or("");
                if f_os == goos && f_arch == goarch {
                    let name = f
                        .get("filename")
                        .and_then(|v| v.as_str())
                        .unwrap_or("")
                        .to_string();
                    return Ok((format!("https://go.dev/dl/{name}"), version));
                }
            }
        }
    }
    bail!("No stable Go archive for {goos}/{goarch}")
}

fn go_bin(ctx: &InstallContext) -> PathBuf {
    let name = if is_windows() { "go.exe" } else { "go" };
    ctx.install_dir.join("bin").join(name)
}

pub struct GoPlugin;

impl Plugin for GoPlugin {
    fn id(&self) -> &'static str {
        "go"
    }

    fn name(&self) -> &'static str {
        "Go"
    }

    fn description(&self) -> &'static str {
        "Download latest stable Go toolchain and set GOROOT / PATH."
    }

    fn status(&self, ctx: &InstallContext) -> crate::plugin::PluginStatus {
        binary_status(
            &go_bin(ctx),
            &ctx.install_dir,
            "Install dir exists but go binary is missing",
        )
    }

    fn install(&self, ctx: &InstallContext) -> Result<InstallResult> {
        let (url, version) = resolve_go_download()?;
        println!("Go {version}");
        println!("URL: {url}");
        let mut progress = download_progress("Downloading Go");
        // Archive root is go/; strip so bin/ lands in install_dir.
        install_archive_from_url(
            &url,
            &ctx.install_dir,
            true,
            None,
            Some(&mut |d, t| progress.update(d, t)),
        )?;
        progress.done();
        let binary = go_bin(ctx);
        if !binary.is_file() {
            bail!("Go extracted but binary not found at {}", binary.display());
        }
        std::fs::write(ctx.install_dir.join(MARKER), format!("{version}\n"))?;
        Ok(InstallResult::new(
            ctx.install_dir.clone(),
            format!("Go {version} installed at {}", ctx.install_dir.display()),
        ))
    }

    fn uninstall(&self, ctx: &InstallContext) -> Result<()> {
        if ctx.install_dir.exists() {
            std::fs::remove_dir_all(&ctx.install_dir)?;
        }
        Ok(())
    }

    /// Same resolver `install` uses, so the string matches the marker it writes.
    fn latest_version(&self, _ctx: &InstallContext) -> Result<Option<String>> {
        Ok(Some(resolve_go_download()?.1))
    }

    fn env_spec(&self, ctx: &InstallContext) -> EnvSpec {
        let root = ctx
            .install_dir
            .canonicalize()
            .unwrap_or_else(|_| ctx.install_dir.clone())
            .display()
            .to_string();
        EnvSpec {
            paths: vec![ctx.install_dir.join("bin")],
            vars: vec![
                ("GOROOT".to_string(), root.clone()),
                ("GO_HOME".to_string(), root),
            ],
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::plugin::InstallState;

    #[test]
    fn go_os_arch_returns_known_pair() {
        let (goos, goarch) = go_os_arch().unwrap();
        assert!(["windows", "darwin", "linux"].contains(&goos));
        assert!(["amd64", "arm64"].contains(&goarch));
    }

    #[test]
    fn status_not_installed_on_empty_dir() {
        let tmp = tempfile::tempdir().unwrap();
        let plugin = GoPlugin;
        let ctx = InstallContext {
            install_dir: tmp.path().join("go"),
            home: tmp.path().to_path_buf(),
            version: None,
            channel: None,
        };
        assert_eq!(plugin.status(&ctx).state, InstallState::NotInstalled);
    }
}
